### PR TITLE
[PHIElimination] Handle subranges in LiveInterval updates

### PR DESCRIPTION
**Pull request by perlfu**
Add subrange tracking and handling for LiveIntervals during PHI elimination.
This requires extending MachineBasicBlock::SplitCriticalEdge to also update subrange intervals.

### CODE REVIEW

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (jayfoad) [inline llvm/lib/CodeGen/PHIElimination.cpp:139]: Normally you would only add this if PHIElimination can do a better job when LiveIntervals is available. If you are just saying that PHIElimination is capable of updating LiveIntervals then you should not add this.

In this case it probably won't make any difference, but in the general case, not adding unnecessary "used if available" dependencies will allow the legacy pass manager to free unneeded analyses sooner, possibly saving some memory.

File: llvm/lib/CodeGen/MachineBasicBlock.cpp
Comment (jayfoad) [inline llvm/lib/CodeGen/MachineBasicBlock.cpp:1309]: Not your fault, but it's a shame that you have to do the `overlaps` check first, because you end up looking up the segment corresponding to StartIndex twice, once in `overlaps` and once in `removeSegment`. It would make more sense to me if `removeSegment` handled the non-overlapping case as a no-op.

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (perlfu) [reply llvm/lib/CodeGen/PHIElimination.cpp]: It seemed to me that if we use an analysis then we should declare that usage.
But I accept your logic and it doesn't make a different in practice here.

File: llvm/lib/CodeGen/MachineBasicBlock.cpp
Comment (perlfu) [reply llvm/lib/CodeGen/MachineBasicBlock.cpp]: I have changed removeSegment to accept these.
I cannot see any obvious reason why it should not, perhaps only to avoid other coding errors?

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (MatzeB) [inline llvm/lib/CodeGen/PHIElimination.cpp:432]: I find this comment highly confusing, and have a hard time seeing when this condition would ever be true... There was a PHI definition `DestReg` here, no? That should mean that every segment here should have been defined at `MBBStartIndex` (where PHI definitions take effect), no? (Except maybe if a subrange was undefined, but then we would have already entered the `isDead()` case above...)

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (MatzeB) [inline llvm/lib/CodeGen/PHIElimination.cpp:440]: as above this should be `==`, no?

**Review by MatzeB — COMMENTED** [review]
* Looks mostly fine
* I have some questions on one `if` condition below.
* Should probably mention in the summary that this adds subrange liveness tracking for critical edge splitting and phi elimination.

File: llvm/lib/CodeGen/LiveInterval.cpp
Comment (perlfu) [reply llvm/lib/CodeGen/LiveInterval.cpp:568]: Done.

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (perlfu) [reply llvm/lib/CodeGen/PHIElimination.cpp:432]: I have moved and reworded this comment for clarity.

There are two details here that matter to understanding this:
* LiveIntervals of PHIs destinations start at the PHI slot index, not the block start.
* Destination copies are not inserted in the order of the PHIs they replace.

Taking the example of the attached test.
Original PHIs look like this:
```
320B      %phi0:vgpr_32 = PHI %s0a:vgpr_32, %bb.0, %s0c:vgpr_32, %bb.1
336B      %phi1:vgpr_32 = PHI %s0b:vgpr_32, %bb.0, %s0d:vgpr_32, %bb.1
```

With intervals as follows:
```
%6 [112r,208r:0) 0@112r  weight:0.000000e+00 -- s0a
%7 [128r,224r:0) 0@128r  weight:0.000000e+00 -- s0b
%8 [208r,256B:0) 0@208r  weight:0.000000e+00 -- s0c
%9 [224r,256B:0) 0@224r  weight:0.000000e+00 -- s0d
%10 [320r,384r:0) 0@320r  weight:0.000000e+00 -- phi0
%11 [336r,384r:0) 0@336r  weight:0.000000e+00 -- phi1
```

phi0 is lowered, giving the following IR:
```
336B      %phi1:vgpr_32 = PHI %s0b:vgpr_32, %bb.0, %s0d:vgpr_32, %bb.1
344B      %phi0:vgpr_32 = COPY %12:vgpr_32
```

phi1 is lowered, giving:
```
312B      %phi1:vgpr_32 = COPY %13:vgpr_32
344B      %phi0:vgpr_32 = COPY %12:vgpr_32
```

Hence COPY for phi0 has index after original PHI, but phi1 has index before original PHI.

**Comment by perlfu** [comment]
Ping

**Comment by perlfu** [comment]
I will merge this on the 9th of November if there is no further feedback.

File: llvm/include/llvm/CodeGen/LiveIntervals.h
Comment (jayfoad) [inline llvm/include/llvm/CodeGen/LiveIntervals.h:142]: Should really have a doxygen comment, even if it's obvious.

**Review by jayfoad — COMMENTED** [review]
Looks good but I will defer to @MatzeB.

File: llvm/lib/CodeGen/PHIElimination.cpp
Comment (arsenm) [inline llvm/lib/CodeGen/PHIElimination.cpp:408]: Can fold this into initializer

Comment (perlfu) [reply llvm/lib/CodeGen/PHIElimination.cpp]: Done

**Review by arsenm — APPROVED** [review]

**Pull request merged**
